### PR TITLE
repo: apt-mark new build-packages as automatically installed

### PR DESCRIPTION
**Pull request by 3v1n0**
This allows to clean them up by just running `apt-get autoremove` once the snapcrafting is done, but in any case it's not breaking anything as it only applies to packages that were not previously installed.

A simpler solution for https://bugs.launchpad.net/snapcraft/+bug/1640143

### CODE REVIEW

**Comment by sergiusens** [comment]
Nice PR, not sure you need `str` when using `format` though

**Comment by sergiusens** [comment]
Can you update the integration tests for this in the base test class by adding a addCleanup of ap autoremove?

**Comment by 3v1n0** [comment]
> Nice PR, not sure you need str when using format though

In fact I was thinking the same (generally the error gets auto-converted), but I followed what it was done few lines above.

**Comment by 3v1n0** [comment]
Ok, tests added...

**Comment by 3v1n0** [comment]
Integration tests updated to...

File: debian/control [hunk 1]
```diff
@@ -15,6 +15,7 @@ Build-Depends: bash-completion,
                python3-lxml,
                python3-magic,
                python3-petname,
+               python3-pexpect,
```
Comment (sergiusens) [inline debian/control]: we don't run the integration tests on package build just in adt, reason for this not being needed

**Comment by 3v1n0** [comment]
@sergiusens ok, removed the debian bit and some double quotes (I've left them in the logger message for consistency to what it was done before)

**Comment by 3v1n0** [comment]
Ah, and do you also want not to use this behavior as default if an env variable (`SNAPCRAFT_IGNORE_APT_MARK_AUTO_INSTALLED` ?) is set?

**Comment by sergiusens** [comment]
no, let's not add magic env vars for this please :-)

File: integration_tests/__init__.py [hunk 2]
```diff
@@ -67,13 +67,42 @@ def run_snapcraft(self, command, project_dir=None, yaml_dir=None):
...
         try:
-            return subprocess.check_output(
+            snapcraft_output = subprocess.check_output(
                 [self.snapcraft_command, '-d'] + command, cwd=cwd,
                 stderr=subprocess.STDOUT, universal_newlines=True)
         except subprocess.CalledProcessError as e:
             self.addDetail('output', content.text_content(e.output))
             raise
 
+        if not os.getenv('SNAPCRAFT_IGNORE_APT_AUTOREMOVE', False):
```
Comment (sergiusens) [inline integration_tests/__init__.py]: can't we just always do the cleanup?

File: integration_tests/__init__.py [hunk 3]
```diff
@@ -67,13 +67,42 @@ def run_snapcraft(self, command, project_dir=None, yaml_dir=None):
...
+                'apt-get autoremove output',
+                content.text_content(autoremove_output.decode('utf-8')))
+        except subprocess.CalledProcessError as e:
+            self.addDetail(
+                'apt-get autoremove error', content.text_content(str(e)))
+            self.addDetail(
+                'apt-get autoremove output',
+                content.text_content(e.output.decode('utf-8')))
+
+            if os.getenv('SNAPCRAFT_APT_AUTOREMOVE_CHECK_FAIL', False):
```
Comment (sergiusens) [inline integration_tests/__init__.py]: why is this needed?

File: snapcraft/internal/repo.py [hunk 4]
```diff
@@ -98,6 +99,13 @@ def install_build_packages(packages):
                                'Dpkg::Progress-Fancy=1',
                                '--no-install-recommends', '-y',
                                'install'] + new_packages, env=env)
+        try:
+                subprocess.check_call(['sudo', 'apt-mark', 'auto'] +
+                                      new_packages, env=env)
```
Comment (elopio) [inline snapcraft/internal/repo.py]: You have four extra spaces here.

File: snapcraft/internal/repo.py [hunk 5]
```diff
@@ -98,6 +99,13 @@ def install_build_packages(packages):
...
+        except subprocess.CalledProcessError as e:
+                logger.warning(
+                    "Impossible to mark packages as auto-installed: {}"
+                    .format(e))
```
Comment (elopio) [inline snapcraft/internal/repo.py]: And here, four extra spaces too.

File: snapcraft/internal/repo.py [hunk 6]
```diff
@@ -98,6 +99,13 @@ def install_build_packages(packages):
                                'Dpkg::Progress-Fancy=1',
                                '--no-install-recommends', '-y',
                                'install'] + new_packages, env=env)
+        try:
+                subprocess.check_call(['sudo', 'apt-mark', 'auto'] +
+                                      new_packages, env=env)
+        except subprocess.CalledProcessError as e:
+                logger.warning(
+                    "Impossible to mark packages as auto-installed: {}"
```
Comment (elopio) [inline snapcraft/internal/repo.py]: Please use single quotes, for consistency :)

**Comment by elopio** [comment]
I'm ok with the tests. Thanks @3v1n0 !

**Comment by 3v1n0** [comment]
@elopio  Fixed the spacing issue, let me know what I should do with the env variables and we should be all set.

**Comment by sergiusens** [comment]
@3v1n0 sorry for causing conflicts, mind rebasing/merging with master?

**Comment by 3v1n0** [comment]
@sergiusens ok, done... And fixed also the wrong `is_dumb_terminal` call in repo.py (it was missing a `()`).
